src/messages/tests/integration.rs

```diff
@@ -0,0 +1,749 @@
+    let signature_share = frost::sign(&signing_package, nonce1[0], &shares[0]).unwrap();
+
+    // valid header
+    let header = create_valid_header(setup.signer1, setup.aggregator);
+
+    let signature = SignatureResponse(signature_share.signature.0.to_bytes());
+    let payload = Payload::SignatureShare(SignatureShare { signature });
+
+    let message = Message {
+        header: header,
+        payload: payload.clone(),
+    };
+
+    // check general structure serialization/deserialization
+    serialize_message(message, setup.signer1, setup.aggregator);
+
+    // check payload serialization/deserialization
+    let mut payload_serialized_bytes = bincode::serialize(&payload).unwrap();
+    // remove the padding. TODO: why is this happening ?
+    payload_serialized_bytes =
+        (&payload_serialized_bytes[4..payload_serialized_bytes.len()]).to_vec();
+
+    // signature is 32 bytes
+    let deserialized_signature: SignatureResponse =
+        bincode::deserialize(&payload_serialized_bytes[0..32]).unwrap();
+    assert_eq!(deserialized_signature, signature);
+}
+
+#[test]
+fn validate_aggregatesignature() {
+    let mut setup = basic_setup();
+
+    // aggregator creates the shares and pubkeys for this round
+    let (shares, pubkeys) =
+        frost::keygen_with_dealer(setup.num_signers, setup.threshold, setup.rng.clone()).unwrap();
+
+    let mut nonces: std::collections::HashMap<u64, Vec<frost::SigningNonces>> =
+        std::collections::HashMap::with_capacity(setup.threshold as usize);
+    let mut commitments: Vec<frost::SigningCommitments> =
+        Vec::with_capacity(setup.threshold as usize);
+
+    // aggregator generates nonces and signing commitments for each participant.
+    for participant_index in 1..(setup.threshold + 1) {
+        let (nonce, commitment) = frost::preprocess(1, participant_index as u64, &mut setup.rng);
+        nonces.insert(participant_index as u64, nonce);
+        commitments.push(commitment[0]);
+    }
+
+    // aggregator generates a signing package
+    let mut signature_shares: Vec<frost::SignatureShare> =
+        Vec::with_capacity(setup.threshold as usize);
+    let message = "message to sign".as_bytes().to_vec();
+    let signing_package = frost::SigningPackage {
+        message: message.clone(),
+        signing_commitments: commitments,
+    };
+
+    // each participant generates their signature share
+    for (participant_index, nonce) in nonces {
+        let share_package = shares
+            .iter()
+            .find(|share| participant_index == share.index)
+            .unwrap();
+        let nonce_to_use = nonce[0];
+        let signature_share = frost::sign(&signing_package, nonce_to_use, share_package).unwrap();
+        signature_shares.push(signature_share);
+    }
+
+    // aggregator generate the final signature
+    let group_signature_res =
+        frost::aggregate(&signing_package, &signature_shares[..], &pubkeys).unwrap();
```

Comment:
Maybe this could also be a utility function, like `full_setup` or something. But it might be too much work :sweat_smile:

Reply:
Yea, lets go in this direction and deduplicate some code. https://github.com/ZcashFoundation/redjubjub/pull/114/commits/8369111861d53f5453e5657c6c6da10205846118